spylib/token.py

```diff
@@ -0,0 +1,428 @@
+            }
+        )
+
+    @classmethod
+    async def new(
+        cls,
+        store_name: str,
+        client_id: str,
+        client_secret: str,
+        code: str,
+    ):
+        """
+        Creates a new instance of the object and requests the token from the API
+        as it requires an async call.
+        """
+        token = cls(store_name=store_name)
+        await token.reset_token(client_id, client_secret, code)
+        return token
+
+    @abstractmethod
+    async def save_token(self):
+        """
+        This method handles saving the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        raise NotImplementedError("Please Implement the save_token method")
+
+    @classmethod
+    @abstractmethod
+    def load_token(self):
+        """
+        This method handles loading the token. By default this does nothing,
+        therefore the developer should override this.
+        """
+        raise NotImplementedError("Please Implement the load_token method.")
+
+    async def _get_token(
+        self, client_id: str, client_secret: str, authorization_code: str
+    ) -> dict:
+        """
+        Makes a request to the access_token endpoint using the client secret,
+        client_id, and the code generated by the OAuth.
+        """
+
+        httpclient = AsyncClient()
+
+        jsondata = {
+            'client_id': client_id,
+            'client_secret': client_secret,
+            'code': authorization_code,
+        }
+
+        response = await httpclient.request(
+            method='post',
+            url=self.oauth_url,
+            json=jsondata,
+            timeout=20.0,
+        )
+        if response.status_code != 200:
+            message = (
+                f'Problem retrieving access token. Status code: {response.status_code} {jsondata}'
+                f'response=> {response.text}'
+            )
+            logger.error(message)
+            raise ValueError(message)
+
+        return response.json()
+
+    async def reset_token(self, client_id: str, client_secret: str, code: str):
+        """Use this function to initialize a new access token for this store"""
+        await self._get_token(client_id, client_secret, code)
+        self.access_token_invalid = False
+
+    @classmethod
+    def get_unique_id(cls) -> str:
+        return ShortUUID().random(length=10)
+
+    # Methods for querying the store
+
+    async def __wait_for_token(self):
+        self.__add_new_tokens()
+        while self.tokens <= 1:
+            await sleep(1)
+            self.__add_new_tokens()
+        self.rest_bucket -= 1
+
+    def __add_new_tokens(self):
+        now = monotonic()
+        time_since_update = now - self.updated_at
+        new_tokens = floor(time_since_update * self.rate)
+        if new_tokens > 1:
+            self.rest_bucket = min(self.rest_bucket + new_tokens, self.rest_bucket_max)
+            self.updated_at = now
```

Comment:
I propose to rename these "tokens" to something like `credits` or maybe `rest_credits` to make the code less confusing. This name always bothered me. Good time to fix this.

Reply:
Should I keep it aligned with the shopify docs or come up with our own names?  I think they refer to them as buckets and the leak for both graphql and rest. But if you have a different name that would be better we can go with that

Reply:
This has been changed to `__await_rest_bucket_refill` and `__fill_rest_bucket` in the latest changes sticking with shopify's naming scheme. Let me know if this is okay!